### PR TITLE
[SPARK-40307][PYTHON] Introduce Arrow-optimized Python UDFs

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Introduce Arrow-optimized Python UDFs. Please refer to [design](https://docs.google.com/document/d/e/2PACX-1vQxFyrMqFM3zhDhKlczrl9ONixk56cVXUwDXK0MMx4Vv2kH3oo-tWYoujhrGbCXTF78CSD2kZtnhnrQ/pub) for design details and micro benchmarks.

There are two ways to enable/disable the Arrow optimization for Python UDFs:

- the Spark configuration `spark.sql.execution.pythonUDF.arrow.enabled`, disabled by default.
- the `useArrow` parameter of the `udf` function, None by default.

The Spark configuration takes effect only when `useArrow` is None. Otherwise, `useArrow` decides whether a specific user-defined function is optimized by Arrow or not.

The reason why we introduce these two ways is to provide both a convenient, per-Spark-session control and a finer-grained, per-UDF control of the Arrow optimization for Python UDFs.

### Why are the changes needed?
Python user-defined function (UDF) enables users to run arbitrary code against PySpark columns. It uses Pickle for (de)serialization and executes row by row.

One major performance bottleneck of Python UDFs is (de)serialization, that is, the data interchanging between the worker JVM and the spawned Python subprocess which actually executes the UDF. 

The PR proposes a better alternative to handle the (de)serialization: Arrow, which is used in the (de)serialization of Pandas UDF already. 

#### Benchmark

The micro benchmarks are conducted in a cluster with 1 driver (i3.2xlarge), 2 workers (i3.2xlarge).  An i3.2xlarge machine has 61 GB Memory, 8 Cores. The datasets used in the benchmarks are generated and sized 5 GB, 10 GB, 20 GB and 40 GB.

As shown below, Arrow-optimized Python UDFs are **~1.4x** faster than non-Arrow-optimized Python UDFs.

image

image

Please refer to [design](https://docs.google.com/document/d/e/2PACX-1vQxFyrMqFM3zhDhKlczrl9ONixk56cVXUwDXK0MMx4Vv2kH3oo-tWYoujhrGbCXTF78CSD2kZtnhnrQ/pub) for details.

### Does this PR introduce _any_ user-facing change?
No, since the Arrow optimization for Python UDFs is disabled by default.

### How was this patch tested?
Unit tests.

Below is the script to generate the result table when the Arrow's type coercion is needed, as in the [docstring](https://github.com/apache/spark/pull/39384/files#diff-2df611ab00519d2d67e5fc20960bd5a6bd76ecd6f7d56cd50d8befd6ce30081bR96-R111) of `_create_py_udf` .
```
import sys
import array
import datetime
from decimal import Decimal

from pyspark.sql import Row
from pyspark.sql.types import *
from pyspark.sql.functions import udf

data = [
    None,
    True,
    1,
    "a",
    datetime.date(1970, 1, 1),
    datetime.datetime(1970, 1, 1, 0, 0),
    1.0,
    array.array("i", [1]),
    [1],
    (1,),
    bytearray([65, 66, 67]),
    Decimal(1),
    {"a": 1},
]

types =  [
    BooleanType(),
    ByteType(),
    ShortType(),
    IntegerType(),
    LongType(),
    StringType(),
    DateType(),
    TimestampType(),
    FloatType(),
    DoubleType(),
    BinaryType(),
    DecimalType(10, 0),
]

df = spark.range(1)
results = []
count = 0
total = len(types) * len(data)
spark.sparkContext.setLogLevel("FATAL")
for t in types:
    result = []
    for v in data:
        try:
            row = df.select(udf(lambda _: v, t)("id")).first()
            ret_str = repr(row[0])
        except Exception:
            ret_str = "X"
        result.append(ret_str)
        progress = "SQL Type: [%s]\n  Python Value: [%s(%s)]\n  Result Python Value: [%s]" % (
            t.simpleString(), str(v), type(v).__name__, ret_str)
        count += 1
        print("%s/%s:\n  %s" % (count, total, progress))
    results.append([t.simpleString()] + list(map(str, result)))

schema = ["SQL Type \\ Python Value(Type)"] + list(map(lambda v: "%s(%s)" % (str(v), type(v).__name__), data))
strings = spark.createDataFrame(results, schema=schema)._jdf.showString(20, 20, False)
print("\n".join(map(lambda line: "    # %s  # noqa" % line, strings.strip().split("\n"))))
```

SPARK-40307

### CODE REVIEW

File: python/pyspark/sql/udf.py [hunk 1]
```diff
@@ -75,6 +81,104 @@ def _create_udf(
...
+            else useArrow
+        )
+
+    regular_udf = _create_udf(f, returnType, evalType)
+    return_type = regular_udf.returnType
+    try:
+        is_func_with_args = len(getfullargspec(f).args) > 0
+    except TypeError:
+        is_func_with_args = False
+    is_output_atomic_type = (
```
Comment (xinrong-meng) [inline python/pyspark/sql/udf.py:134]: The limitation ought to be eliminated as a follow-up.

File: python/pyspark/sql/udf.py [hunk 2]
```diff
@@ -75,6 +81,104 @@ def _create_udf(
...
+    # with/without Arrow optimization. That's the main reason the Arrow optimization for Python
+    # UDFs is disabled by default.
+    # +-----------------------------+--------------+----------+------+------+----------------+-----------------------------+----------+----------------------+---------+-----------+----------------------------+----------+--------------+  # noqa
+    # |SQL Type \ Python Value(Type)|None(NoneType)|True(bool)|1(int)|a(str)|1970-01-01(date)|1970-01-01 00:00:00(datetime)|1.0(float)|array('i', [1])(array)|[1](list)|(1,)(tuple)|bytearray(b'ABC')(bytearray)|1(Decimal)|{'a': 1}(dict)|  # noqa
+    # +-----------------------------+--------------+----------+------+------+----------------+-----------------------------+----------+----------------------+---------+-----------+----------------------------+----------+--------------+  # noqa
+    # |                      boolean|             X|         X|     X|     X|               X|                            X|         X|                     X|        X|          X|                           X|         X|             X|  # noqa
```
Comment (HyukjinKwon) [inline python/pyspark/sql/udf.py:99]: Why is it all `X`? X means an error.

File: python/pyspark/sql/tests/test_arrow_python_udf.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                udf(lambda x: str(x))("map"),
+            )
+            .first()
+        )
+
+        # The input is NumPy array when the optimization is on.
+        self.assertEquals(row[0], "[1 2 3]")
+        self.assertEquals(row[1], "{'a': 'b'}")
+
+    def test_useArrow(self):
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_arrow_python_udf.py:80]: Let's name it `test_use_arrow`

**Review by HyukjinKwon — APPROVED** [review]
LGTM otherwise

**Review by zhengruifeng — APPROVED** [review]

File: python/pyspark/sql/udf.py (hunk 2 above)
Comment (xinrong-meng) [reply python/pyspark/sql/udf.py]: Good catch! I re-generated the table and added a note for the library versions used.

```
$ conda list | grep -e 'python\|pyarrow\|pandas'
pandas                    1.5.2                    pypi_0    pypi
pandas-stubs              1.2.0.53                 pypi_0    pypi
pyarrow                   10.0.1                   pypi_0    pypi
python                    3.9.15               h218abb5_2  
python-dateutil           2.8.2                    pypi_0    pypi
```

File: python/pyspark/sql/tests/test_arrow_python_udf.py (hunk 3 above)
Comment (xinrong-meng) [reply python/pyspark/sql/tests/test_arrow_python_udf.py]: Sounds good! Renamed.

**Comment by xinrong-meng** [comment]
Test failures in CI jobs are irrelevant to the PR. I'll rebase the PR to the latest master.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by xinrong-meng** [comment]
Thank you both! Merged to master.

File: python/pyspark/sql/tests/test_udf.py [hunk 4]
```diff
@@ -804,6 +804,54 @@ def test_udf_with_rand(self):
...
 
 
+class UDFTests(BaseUDFTests, ReusedSQLTestCase):
+    @classmethod
+    def setUpClass(cls):
+        super(BaseUDFTests, cls).setUpClass()
+        cls.spark.conf.set("spark.sql.execution.pythonUDF.arrow.enabled", "false")
+
+
+def test_use_arrow(self):
```
Comment (ueshin) [inline python/pyspark/sql/tests/test_udf.py:814]: @xinrong-meng Who runs this test?

Comment (xinrong-meng) [reply python/pyspark/sql/tests/test_udf.py:814]: Good catch! Removed it https://github.com/apache/spark/pull/40725/commits/63ef94e6855a9b9cbf7f20282f28cebaf06cd3c6. `test_use_arrow` is duplicated in PythonUDFArrowTestsMixin of test_arrow_python_udf.py.

**Pull request closed without merging**